Adding support for metric publishing, manually tested and tests pending.

c2tarun opened the pull request:
*Issue #, if available:*

*Description of changes:*

*Description of how you validated changes:*

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected

*Examples?*

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mndeveci commented on samtranslator/metrics/metrics.py:
```diff
@@ -0,0 +1,156 @@
+        try:
+            self.cloudwatch_client.put_metric_data(Namespace=namespace, MetricData=metric_data)
+        except Exception as e:
+            LOG.error("Failed to report {} metrics".format(len(metric_data)))
+            LOG.exception(e)
```
nit: I think there is a parameter in the logging framework to pass the exception rather than doing it in 2 lines

c2tarun replied:
Fixed!!

mndeveci commented on samtranslator/metrics/metrics.py:
```diff
@@ -0,0 +1,156 @@
+
+    def publish(self, namespace, metrics):
+        """Method to actually publish them metric"""
+        LOG.info("Dummy publisher ignoring {} metrices".format(len(metrics)))
```
Can we do nothing or change this to `debug`? My concern is coming from people who is using this as a dependency in their library, for instance we might see this message from time to time in SAM CLI too.

c2tarun replied:
I'll change this to `debug`.

mndeveci commented on samtranslator/metrics/metrics.py:
```diff
@@ -0,0 +1,156 @@
+        """
+        self.metrics_cache.append(MetricDatum(name, value, unit, dimensions))
+
+    def record_count(self, name, value, unit=Unit.Count, dimensions=[]):
+        """
+        Create metric with unit Count.
+
+        :param name: metric name
+        :param value: value of metric
+        :param unit: unit of metric (try using values from Unit class)
+        :param dimensions: array of dimensions applied to the metric
+        """
+        self._record_metric(name, value, unit, dimensions)
+
+    def record_latency(self, name, value, unit=Unit.Milliseconds, dimensions=[]):
```
Given that these methods are specialized, can we make them easier by removing `unit` and `dimensions` parameters?

c2tarun replied:
I'll remove `unit` we kind of need `dimensions`

c2tarun commented on samtranslator/metrics/metrics.py:
```diff
@@ -0,0 +1,156 @@
+import logging
```
Done

mndeveci commented on samtranslator/metrics/metrics.py:
```diff
@@ -0,0 +1,158 @@
+            LOG.exception("Failed to report {} metrics".format(len(metric_data)), exc_info=e)
+
+
+class DummyMetricsPublisher(MetricsPublisher):
+    def __init__(self):
+        MetricsPublisher.__init__(self)
+
+    def publish(self, namespace, metrics):
+        """Method to actually publish them metric"""
+        LOG.debug("Dummy publisher ignoring {} metrices".format(len(metrics)))
```
Sorry I didn't catch this in the first round but can we add some documentation that this won't publish anything?
And also the doc on the `publish` method seems to be suggesting that this method will publish the metrics.

c2tarun replied:
Addressed, please let me know if it makes sense now.

mndeveci reviewed (commented):
Thanks for addressing the feedbacks @c2tarun , left one comment about docs, and I was wondering if we need `tests/metrics/__init__.py `?

c2tarun commented:
> Thanks for addressing the feedbacks @c2tarun , left one comment about docs, and I was wondering if we need `tests/metrics/__init__.py `?

You are right, we don't, good catch :) remove it.
